static/js/util/profile_edit.js

```diff
@@ -220,6 +223,106 @@ export function boundCheckbox(keySet: string[], label: string|React$Element<*>):
   );
 }
 
+const hasValue = R.and(
+  R.compose(R.not, R.isEmpty),
+  R.compose(R.not, R.isNil),
+);
+
+const getComponentForType = (components: GoogleMapsAddressComponent[], type: string) => (
+  R.find(R.propSatisfies(R.contains(type), "types"), components)
+);
+
+const profileFields = (addressMapping: AddressComponentKeyMapping, components: GoogleMapsAddressComponent[]) => {
+  let profile = {};
+  _.forOwn(addressMapping, (keySet, gmapType) => {
+    const component = getComponentForType(components, gmapType);
+    _.set(profile, keySet, component.long_name);
+  });
+  return profile;
+};
+
+export function boundGeosuggest(
+  addressMapping: AddressComponentKeyMapping,
+  label: string|React$Element<*>,
+  { id, placeholder, types }: { id: string, placeholder: string, types: string[] } = {}
+): React$Element<*> {
+  const {
+    profile,
+    errors,
+    updateProfile,
+    validator,
+    updateValidationVisibility,
+  } = this.props;
+
+  const updateMultipleValidationVisibility = (addressMapping) => {
+    R.values(addressMapping).forEach((keySet) => (
+      updateValidationVisibility(keySet)
+    ));
+  };
+
+  const onSuggestSelect = (suggest) => {
+    if (!suggest || !suggest.gmaps || !suggest.gmaps.address_components) {
+      return;
+    }
+
+    const newProfile = R.merge(
+      R.clone(profile),
+      profileFields(addressMapping, suggest.gmaps.address_components)
+    );
+    updateMultipleValidationVisibility(addressMapping);
+    updateProfile(newProfile, validator);
+  };
+
+  const onBlur = (value) => {
+    if (!value) {
+      let clone = _.cloneDeep(profile);
+      _.forOwn(addressMapping, (keySet, gmapType) => {  // eslint-disable-line no-unused-vars
+        _.set(clone, keySet, null);
+      });
+      updateMultipleValidationVisibility(addressMapping);
+      updateProfile(clone, validator);
+    }
+  };
+
+  const hasAddressProps = (profile) => (
+    R.all(R.pathSatisfies(hasValue, R.__, profile), R.values(addressMapping))
+  );
+
+  const hasExistingAddress = R.allPass([
+    hasValue,
+    hasAddressProps,
+  ]);
+
+  const formatInitialAddress = R.ifElse(
+    hasExistingAddress,
+    R.compose(R.join(", "), R.props(R.values(addressMapping))),
+    R.always(""),
+  );
+
+  const initial = formatInitialAddress(profile);
+
+  const hasErrors = (errors) => (
+    R.any(R.pathSatisfies(hasValue, R.__, errors), R.values(addressMapping))
+  );
```

Comment:
this is pretty similar to `hasAddressProps` above, maybe the two could be rewritten to reuse some code, something like

```js
const _addressPropCheck = R.flip(R.pathSatisifes(hasValue, R.__));

const addressPropCheck = checkFunc => (
  R.compose(checkFunc(R.__, R.values(addressMapping)), _addressPropCheck)
);

const hasAddressProps = addressPropCheck(R.all);

const hasErrors = addressPropCheck(R.any);
```

I think that works! haha.